sw/host/opentitanlib/src/util/present.rs

```diff
@@ -0,0 +1,274 @@
+    }
+
+    /// Encrypt a 64-bit block.
+    pub fn encrypt_block(&self, block: u64) -> u64 {
+        let mut state = block;
+        state ^= self.round_keys[0];
+        for round_key in &self.round_keys[1..] {
+            state = s_box_layer(state);
+            state = p_box_layer(state);
+            state ^= round_key;
+        }
+        state
+    }
+
+    /// Decrypt a 64-bit block.
+    pub fn decrypt_block(&self, block: u64) -> u64 {
+        let mut state = block;
+        for round_key in self.round_keys[1..].iter().rev() {
+            state ^= round_key;
+            state = p_box_layer_dec(state);
+            state = s_box_layer_dec(state);
+        }
+        state ^ self.round_keys[0]
+    }
+}
+
+const S_BOX: [u8; 16] = [
+    0x0c, 0x05, 0x06, 0x0b, 0x09, 0x00, 0x0a, 0x0d, 0x03, 0x0e, 0x0f, 0x08, 0x04, 0x07, 0x01, 0x02,
+];
+
+const S_BOX_INV: [u8; 16] = [
+    0x05, 0x0e, 0x0f, 0x08, 0x0c, 0x01, 0x02, 0x0d, 0x0b, 0x04, 0x06, 0x03, 0x00, 0x07, 0x09, 0x0a,
+];
+
+const P_BOX: [u8; 64] = [
+    0x00, 0x10, 0x20, 0x30, 0x01, 0x11, 0x21, 0x31, 0x02, 0x12, 0x22, 0x32, 0x03, 0x13, 0x23, 0x33,
+    0x04, 0x14, 0x24, 0x34, 0x05, 0x15, 0x25, 0x35, 0x06, 0x16, 0x26, 0x36, 0x07, 0x17, 0x27, 0x37,
+    0x08, 0x18, 0x28, 0x38, 0x09, 0x19, 0x29, 0x39, 0x0a, 0x1a, 0x2a, 0x3a, 0x0b, 0x1b, 0x2b, 0x3b,
+    0x0c, 0x1c, 0x2c, 0x3c, 0x0d, 0x1d, 0x2d, 0x3d, 0x0e, 0x1e, 0x2e, 0x3e, 0x0f, 0x1f, 0x2f, 0x3f,
+];
+
+const P_BOX_INV: [u8; 64] = [
+    0x00, 0x04, 0x08, 0x0c, 0x10, 0x14, 0x18, 0x1c, 0x20, 0x24, 0x28, 0x2c, 0x30, 0x34, 0x38, 0x3c,
+    0x01, 0x05, 0x09, 0x0d, 0x11, 0x15, 0x19, 0x1d, 0x21, 0x25, 0x29, 0x2d, 0x31, 0x35, 0x39, 0x3d,
+    0x02, 0x06, 0x0a, 0x0e, 0x12, 0x16, 0x1a, 0x1e, 0x22, 0x26, 0x2a, 0x2e, 0x32, 0x36, 0x3a, 0x3e,
+    0x03, 0x07, 0x0b, 0x0f, 0x13, 0x17, 0x1b, 0x1f, 0x23, 0x27, 0x2b, 0x2f, 0x33, 0x37, 0x3b, 0x3f,
+];
+
+/// Generate the round_keys for an 80-bit key.
+fn generate_round_keys_80(key: Vec<u8>, rounds: usize) -> Vec<u64> {
+    // Pad out key so it fits in a u128 later.
+    let mut orig_key = key;
+    let mut key = vec![0u8; 6];
+    key.append(&mut orig_key);
+
+    let mut round_keys = Vec::new();
+    for i in 1..rounds + 1 {
+        // Convert key into a u128 for easier bit manipulation.
+        let next_key: &[u8; 16] = key.as_slice().try_into().unwrap();
+        let mut next_key = u128::from_le_bytes(*next_key);
+
+        // rawKey[0:64]
+        let round_key = (next_key >> 16) as u64;
+
+        // 1. Rotate bits
+        // rawKey[19:len(rawKey)]+rawKey[0:19]
+        next_key = (next_key & 0x7ffff) << 61 | next_key >> 19;
+
+        // 2. SBox
+        // rawKey[76:80] = S(rawKey[76:80])
+        next_key =
+            (S_BOX[(next_key >> 76) as usize] as u128) << 76 | (next_key & !0u128 >> (128 - 76));
+
+        // 3. Salt
+        // rawKey[15:20] ^ i
+        next_key ^= (i as u128) << 15;
+
+        round_keys.push(round_key);
+        key = next_key.as_bytes().to_vec();
```

Comment:
You can drop this line: you're doing the operation in place so key == next_key already.